lib/c-api/src/lib.hh

```diff
@@ -23,6 +23,48 @@ struct clingo_lib {
 
 namespace CppClingo::CAPI {
 
+//! Helper to manage C handlers with userdata.
+//!
+//! The handlers are expected to have a `free` function pointer to cleanup the
+//! userdata.
+template <class Handler> class CHandler {
+  public:
+    CHandler(Handler const *handler, void *data) noexcept
+        : handler_{handler != nullptr ? std::optional{*handler} : std::nullopt}, data_{data} {}
+
+    CHandler(const CHandler &other) = delete;
+    auto operator=(const CHandler &other) -> CHandler & = delete;
+
+    CHandler(CHandler &&other) noexcept
+        : handler_{std::exchange(other.handler_, std::nullopt)}, data_{std::exchange(other.data_, nullptr)} {}
+    auto operator=(CHandler &&other) noexcept -> CHandler & {
+        if (this != &other) {
+            cleanup();
+            handler_ = std::exchange(other.handler_, std::nullopt);
+            data_ = std::exchange(other.data_, nullptr);
+        }
+        return *this;
+    }
+
+    auto operator->() -> Handler * { return handler(); }
+    operator bool() const noexcept { return handler_.has_value(); }
+
+    auto handler() -> Handler * { return handler_ ? &*handler_ : nullptr; }
+    auto data() -> void * { return data_; }
+
+    ~CHandler() { cleanup(); }
+
+  private:
+    void cleanup() noexcept {
+        if (handler_ && handler_->free) {
+            handler_->free(data_);
+        }
+    }
+
+    std::optional<Handler> handler_;
```

Comment:
What's the motivation for using an optional instead of directly using `nullptr` as the "none" value?

Reply:
The C struct is stored by value not pointer. I thought to copy it to avoid any lifetime issues here.